# RAPAKA MANOJ KUMAR

## AI ELITE 13 (243)

### weather_data_analysis_assignment1

#### Importing Libraries

In [1]:
import pandas as pd
import numpy as np

#### Reading the csv file

In [2]:
df_weather = pd.read_csv('data/weather_data (1).csv')

### question no 1

### Q1) What is the shape of given dataset ?

a - (145459, 22)

b - (145460, 23) ✅ #correct

c - (23, 145460)

d - (22, 145459)

Answer - (145460,23)

In [3]:
df_weather.shape

(145460, 23)

### Q2 ) How many integer columns are given in the dataset?

a - 3

b - 4

c - 2

d - 0 ✅ #correct

Answer - 0

In [4]:
int_col = df_weather.select_dtypes(include = [ 'int64']).columns
print( int_col )               #count = 0

Index([], dtype='object')


In [5]:
#for dual confirmation check the dtype in each column by seeing
df_weather.head()

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W           44.0          W  ...        71.0         22.0   
1         WNW           44.0        NNW  ...        44.0         25.0   
2         WSW           46.0          W  ...        38.0         30.0   
3          NE           24.0         SE  ...        45.0         16.0   
4           W           41.0        ENE  ...        82.0         33.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  

[5 rows x 23 columns]

### Q2) What is the datatype of ' WindGustSpeed' column?

a - object

b - int64

c - float64 ✅ #correct

d - datetime

Answer - float64

In [6]:
df_weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

In [7]:
print("Data Type of each columns : \n", df_weather.dtypes)    #WindGustSpeed    float64

Data Type of each columns : 
 Date              object
Location          object
MinTemp          float64
MaxTemp          float64
Rainfall         float64
Evaporation      float64
Sunshine         float64
WindGustDir       object
WindGustSpeed    float64
WindDir9am        object
WindDir3pm        object
WindSpeed9am     float64
WindSpeed3pm     float64
Humidity9am      float64
Humidity3pm      float64
Pressure9am      float64
Pressure3pm      float64
Cloud9am         float64
Cloud3pm         float64
Temp9am          float64
Temp3pm          float64
RainToday         object
RainTomorrow      object
dtype: object


### Q3) How many missing values are there in 'Sunshine' column?

a)75625

b)82670

c)135197

d)69835 ✅

Answer - 69835

In [8]:
df_weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

In [9]:
Total_entries  = 145460  
non_null_sunshine = 75625    #total non - null entries in column Sunshine col

null_val_sunshine = (Total_entries - non_null_sunshine)
print(" total null entries in column Sunshine col : \n" , null_val_sunshine)

 total null entries in column Sunshine col : 
 69835


In [10]:
df_weather['Sunshine'].isnull().sum()

69835

### Q4) How many non-null values there in ' Humidity9am' column?

a)142806

b)2654 ✅

c)141232

d)140953

Answer - 142806

In [11]:
df_weather['Humidity9am'].notnull().sum()

142806

In [12]:
Total_entries  = 145460  
non_null_Humidity9am = 142806    #total non - null entries in column Humidity9am col

null_val_Humidity9am = (Total_entries - non_null_Humidity9am)
print(" total null entries in column Humidity9am col : \n" , null_val_Humidity9am)

 total null entries in column Humidity9am col : 
 2654


### Q5) How many object columns are there in the given dataset?

a)4

b)5

c)6

d)7 ✅

Answer - 7

In [13]:
df_weather.info()                #dtypes: float64(16), object(7)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

In [14]:
df_weather.select_dtypes(include='object')

Date Location WindGustDir WindDir9am WindDir3pm RainToday  \
0       2008-12-01   Albury           W          W        WNW        No   
1       2008-12-02   Albury         WNW        NNW        WSW        No   
2       2008-12-03   Albury         WSW          W        WSW        No   
3       2008-12-04   Albury          NE         SE          E        No   
4       2008-12-05   Albury           W        ENE         NW        No   
...            ...      ...         ...        ...        ...       ...   
145455  2017-06-21    Uluru           E         SE        ENE        No   
145456  2017-06-22    Uluru         NNW         SE          N        No   
145457  2017-06-23    Uluru           N         SE        WNW        No   
145458  2017-06-24    Uluru          SE        SSE          N        No   
145459  2017-06-25    Uluru         NaN        ESE        ESE        No   

       RainTomorrow  
0                No  
1                No  
2                No  
3                No  
4                No  
...             ...  
145455           No  
145456           No  
145457           No  
145458           No  
145459          NaN  

[145460 rows x 7 columns]

### Q6) How much memory is occupied by the dataset on your computers RAM?

a)Less than 25 MB

b)More than 25 MB ✅ # its more than 25 MegaBytes

c)More than 25 Mb ✅ # on the other hand its also more than 25 Megabits

d)Equal to 25.5 Mb

Answer - More than 25 MB ✅
More than 25 Mb ✅

In [15]:
df_weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

In [19]:
df_weather.memory_usage().sum()

26764768

In [20]:
df_weather.columns

Index(['Date', 'Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation',
       'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm',
       'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
       'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
       'Temp3pm', 'RainToday', 'RainTomorrow'],
      dtype='object')

### Q 7 ) What is the percentage of values missing in 'MinTemp' column?

a)0.866905

b)1.020899 ✅

c)43.166506

d)7.098859

Answer - 1.02

In [21]:
Total_entries  = 145460  
non_null_MinTemp = 143975    #total non - null entries in column MinTemp col

percent_missing = ( (Total_entries - non_null_MinTemp ) / Total_entries ) * 100
print("  percentage of values missing in 'MinTemp' column : \n" , percent_missing)

  percentage of values missing in 'MinTemp' column : 
 1.0208992162793895


In [23]:
percentage_missing = (df_weather.isnull().sum()*100/len(df_weather))

In [24]:
percentage_missing['MinTemp']

1.0208992162793895

### Q 8 ) Did it rain in 'Uluru' location on '23rd June 2017' ? #23-06-2017

Yes

No ✅

Answer - No

In [26]:
df_weather.query('Location == "Uluru" and Date == "2017-06-23"')

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
145457  2017-06-23    Uluru      5.4     26.9       0.0          NaN   

        Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  \
145457       NaN           N           37.0         SE  ...        53.0   

        Humidity3pm  Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  \
145457         24.0       1021.0       1016.8       NaN       NaN     12.5   

        Temp3pm  RainToday  RainTomorrow  
145457     26.1         No            No  

[1 rows x 23 columns]

In [28]:
df_weather[['Date','Location','Rainfall','RainToday']].iloc[145457]

Date         2017-06-23
Location          Uluru
Rainfall            0.0
RainToday            No
Name: 145457, dtype: object

In [29]:
df_weather.tail(3)         #check 'Rainfall'[5th] and 'RainToday' [last 2nd col]

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
145457  2017-06-23    Uluru      5.4     26.9       0.0          NaN   
145458  2017-06-24    Uluru      7.8     27.0       0.0          NaN   
145459  2017-06-25    Uluru     14.9      NaN       0.0          NaN   

        Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  \
145457       NaN           N           37.0         SE  ...        53.0   
145458       NaN          SE           28.0        SSE  ...        51.0   
145459       NaN         NaN            NaN        ESE  ...        62.0   

        Humidity3pm  Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  \
145457         24.0       1021.0       1016.8       NaN       NaN     12.5   
145458         24.0       1019.4       1016.5       3.0       2.0     15.1   
145459         36.0       1020.2       1017.9       8.0       8.0     15.0   

        Temp3pm  RainToday  RainTomorrow  
145457     26.1         No            No  
145458     26.0         No            No  
145459     20.9         No           NaN  

[3 rows x 23 columns]

### Q9 ) What was the temperature at 3 PM in 'Albury' location on '3rd Dec 2008' ?

24.3

23.2 ✅

26.5

21.0

Answer - 23.2

In [32]:
df_weather[['Date','Location','Temp3pm']].iloc[2]

Date        2008-12-03
Location        Albury
Temp3pm           23.2
Name: 2, dtype: object

In [33]:
df_weather[[ 'Date' , 'Location' , 'Temp3pm' ]].head(3)       #check last data point(row)

Date Location  Temp3pm
0  2008-12-01   Albury     21.8
1  2008-12-02   Albury     24.3
2  2008-12-03   Albury     23.2

### Q10 ) Which of the following is correct?

a)In Albury on 4th Dec 2008, Wind Direction at 3 PM was observed to change from SE (at 9 AM) to E. ✅

b)In Albury on 4th Dec 2008, Wind Direction at 3 PM was observed to change from W (at 9 AM) to E.

c)In Albury on 4th Dec 2008, Wind Direction at 3 PM was observed to change from E (at 9 AM) to SE.

d)In Albury on 5th Dec 2008, Wind Direction at 3 PM was observed as NW. ✅

Answer - a and d

In [34]:
df_weather.columns

Index(['Date', 'Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation',
       'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm',
       'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
       'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
       'Temp3pm', 'RainToday', 'RainTomorrow'],
      dtype='object')

In [35]:
df_weather[[ 'Date' , 'Location' , 'WindDir9am' , 'WindDir3pm' ]].head(5)     #analysis

Date Location WindDir9am WindDir3pm
0  2008-12-01   Albury          W        WNW
1  2008-12-02   Albury        NNW        WSW
2  2008-12-03   Albury          W        WSW
3  2008-12-04   Albury         SE          E
4  2008-12-05   Albury        ENE         NW

In [37]:
df_weather.query('Date == "2008-12-04" and Location == "Albury"')

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
3          NE           24.0         SE  ...        45.0         16.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   

   RainTomorrow  
3            No  

[1 rows x 23 columns]

In [39]:
df_weather[['Date', 'Location', 'WindDir9am', 'WindDir3pm']].iloc[3]

Date          2008-12-04
Location          Albury
WindDir9am            SE
WindDir3pm             E
Name: 3, dtype: object

### Q11) Out of total 145460 days, on how many days it rained?

110319

142199

31880 ✅

142193

Answer - 31880

In [40]:
a = df_weather['RainToday']
list1 = [i if i == 'Yes' else ... for i in a]     #List Comprehension + Ternary 
print(" Total count of Rained days are : \n", list1.count('Yes'))

 Total count of Rained days are : 
 31880


In [41]:
df_weather['RainToday'].head(10)

0     No
1     No
2     No
3     No
4     No
5     No
6     No
7     No
8     No
9    Yes
Name: RainToday, dtype: object

In [43]:
df_weather.query('RainToday == "Yes"')['RainToday'].count()

31880

### 12.Choose all the correct options:

Minimum MaxTemp observed is -4.80 ✅

25th Percentile of Wind Speed at 9 AM is 7.00 ✅

Median Temperature observed at 3 PM was 21.10 ✅

On average 5.468232 evaporation is recorded. ✅

None of the above

Answer - ALL Correct

In [44]:
#a)
df_weather['MaxTemp'].min()

-4.8

In [45]:
#B)
df_weather.describe()

MinTemp        MaxTemp       Rainfall   Evaporation  \
count  143975.000000  144199.000000  142199.000000  82670.000000   
mean       12.194034      23.221348       2.360918      5.468232   
std         6.398495       7.119049       8.478060      4.193704   
min        -8.500000      -4.800000       0.000000      0.000000   
25%         7.600000      17.900000       0.000000      2.600000   
50%        12.000000      22.600000       0.000000      4.800000   
75%        16.900000      28.200000       0.800000      7.400000   
max        33.900000      48.100000     371.000000    145.000000   

           Sunshine  WindGustSpeed   WindSpeed9am   WindSpeed3pm  \
count  75625.000000  135197.000000  143693.000000  142398.000000   
mean       7.611178      40.035230      14.043426      18.662657   
std        3.785483      13.607062       8.915375       8.809800   
min        0.000000       6.000000       0.000000       0.000000   
25%        4.800000      31.000000       7.000000      13.000000   
50%        8.400000      39.000000      13.000000      19.000000   
75%       10.600000      48.000000      19.000000      24.000000   
max       14.500000     135.000000     130.000000      87.000000   

         Humidity9am    Humidity3pm   Pressure9am    Pressure3pm  \
count  142806.000000  140953.000000  130395.00000  130432.000000   
mean       68.880831      51.539116    1017.64994    1015.255889   
std        19.029164      20.795902       7.10653       7.037414   
min         0.000000       0.000000     980.50000     977.100000   
25%        57.000000      37.000000    1012.90000    1010.400000   
50%        70.000000      52.000000    1017.60000    1015.200000   
75%        83.000000      66.000000    1022.40000    1020.000000   
max       100.000000     100.000000    1041.00000    1039.600000   

           Cloud9am      Cloud3pm        Temp9am       Temp3pm  
count  89572.000000  86102.000000  143693.000000  141851.00000  
mean       4.447461      4.509930      16.990631      21.68339  
std        2.887159      2.720357       6.488753       6.93665  
min        0.000000      0.000000      -7.200000      -5.40000  
25%        1.000000      2.000000      12.300000      16.60000  
50%        5.000000      5.000000      16.700000      21.10000  
75%        7.000000      7.000000      21.600000      26.40000  
max        9.000000      9.000000      40.200000      46.70000

In [46]:
#C)
# float_col = df_weather.select_dtypes(include = [ 'float64']).columns
df_weather['Temp3pm'].median()

21.1

In [47]:
#D)
#Check the total count in evaporation in describe (i.e. 1st row and evaporation column ,  you will get:)
count = 82670.000000
print("Count of Evaporation col : \n", count)
print("Sum of Evaporation col : \n", df_weather['Evaporation'].sum())
print("On average 5.468232 evaporation is recorded: \n", (df_weather['Evaporation'].sum() / count))

Count of Evaporation col : 
 82670.0
Sum of Evaporation col : 
 452058.69999999995
On average 5.468232 evaporation is recorded: 
 5.468231522922462


In [49]:
# Minimum MaxTemp observed is -4.80

df_weather['MaxTemp'].min()

-4.8

In [51]:
# 25th Percentile of Wind Speed at 9 AM is 7.00

df_weather['WindSpeed9am'].quantile(0.25)

7.0

In [52]:
# Median Temperature observed at 3 PM was 21.10

df_weather['Temp3pm'].median()

21.1

In [54]:
# On average 5.468232 evaporation is recorded.

df_weather['Evaporation'].mean()

5.468231522922462

### Q13) Variance in rainfall recorded is?



71.00

71.88 ✅

71.44

71.22

Answer - 71.88

In [55]:
df_weather['Rainfall'].var()

71.87749691637731

### 14.Which Australian Town/City data is not available in the given dataset?

GoldCoast

Albany

Perth

Orange ✅

Melbourne

Answer - orange

In [56]:
#way1
abc = ["GoldCoast","Albany","Perth","Orange","Melbourne"]
print("Missing Australian Town/City data is not available in the given dataset: \n")
for i in abc:
    if i in df_weather['Location'].value_counts():
        pass
    else:
        print(i)

Missing Australian Town/City data is not available in the given dataset: 

Orange


In [57]:
#way2
abc = ["GoldCoast","Albany","Perth","Orange","Melbourne"]
print("Missing Australian Town/City data is not available in the given dataset: \n")
for i in abc:
    if i not in df_weather['Location'].value_counts():
        print(i)

Missing Australian Town/City data is not available in the given dataset: 

Orange


In [58]:
df_weather['Location'].unique()

array(['Albury', 'BadgerysCreek', 'Cobar', 'CoffsHarbour', 'Moree',
       'Newcastle', 'NorahHead', 'NorfolkIsland', 'Penrith', 'Richmond',
       'Sydney', 'SydneyAirport', 'WaggaWagga', 'Williamtown',
       'Wollongong', 'Canberra', 'Tuggeranong', 'MountGinini', 'Ballarat',
       'Bendigo', 'Sale', 'MelbourneAirport', 'Melbourne', 'Mildura',
       'Nhil', 'Portland', 'Watsonia', 'Dartmoor', 'Brisbane', 'Cairns',
       'GoldCoast', 'Townsville', 'Adelaide', 'MountGambier', 'Nuriootpa',
       'Woomera', 'Albany', 'Witchcliffe', 'PearceRAAF', 'PerthAirport',
       'Perth', 'SalmonGums', 'Walpole', 'Hobart', 'Launceston',
       'AliceSprings', 'Darwin', 'Katherine', 'Uluru'], dtype=object)

In [59]:
cities = list(df_weather['Location'].unique())
desired_cities = ['GoldCoast','Albany','Perth','Orange','Melbourne']
missing_city = [city for city in desired_cities if city not in cities]
print(missing_city)

['Orange']
